src/backend/distributed/shared_library_init.c

```diff
@@ -357,7 +382,56 @@ _PG_init(void)
 	{
 		DoInitialCleanup();
 	}
-	columnar_init();
+
+	/* ensure columnar module is loaded at the right time */
+	load_file(COLUMNAR_MODULE_NAME, false);
+
+	/*
+	 * Now, acquire symbols from columnar module. First, acquire
+	 * the address of the set options hook, and set it so that we
+	 * can propagate options changes.
+	 */
+	ColumnarTableSetOptions_hook_type **ColumnarTableSetOptions_hook_ptr =
+		(ColumnarTableSetOptions_hook_type **) find_rendezvous_variable(
+			"ColumnarTableSetOptions_hook");
+
+	/* rendezvous variable registered during columnar initialization */
+	Assert(ColumnarTableSetOptions_hook_ptr != NULL);
+	Assert(*ColumnarTableSetOptions_hook_ptr != NULL);
+
+	**ColumnarTableSetOptions_hook_ptr = ColumnarTableSetOptionsHook;
+
+	/*
+	 * Acquire symbols for columnar functions that citus calls.
+	 */
+	void *handle = NULL;
+	ExternColumnarSupportsIndexAM = (ColumnarSupportsIndexAM_type) (void *)
+									load_external_function(COLUMNAR_MODULE_NAME,
+														   "ColumnarSupportsIndexAM",
+														   true, &handle);
```

Comment:
I somewhat like the define below a bit better compared to loading external functions here inline.
Could we reuse the `INIT_COLUMNAR_PASSTHROUGH_SYMBOL` here? I guess not, as those are specifically loading UDF's from columnar.

Adding a separate define might be ok here. Don't know if that makes the whole thing more or less readable though.

What is the reason we `load_extern_function` here, and load the UDF's via a define?

Reply:
I refactored a bit to improve these macros and use them for all the initialization (except the first one) -- still not perfect, but a little cleaner. 

The reason I did things a little differently for the UDF passthroughs is because I want the UDFs to have static symbols (the wrapper function) that call a function pointer. If the symbol was actually just a function _pointer_, then dlsym (when the UDF is called) would return the address of that variable rather than the function.

The other symbols we're loading don't have that problem; the citus code can call a function pointer just fine. So I didn't see the need for wrapper functions. After an upgrade script to load the symbols from the columnar module instead, we'll remove the wrapper symbols anyway.